runtime-light/tl/tl-core.h

```diff
@@ -0,0 +1,101 @@
+// Compiler for PHP (aka KPHP)
+// Copyright (c) 2024 LLC «V Kontakte»
+// Distributed under the GPL v3 License, see LICENSE.notice.txt
+
+#pragma once
+
+#include <cstddef>
+#include <optional>
+#include <string_view>
+#include <type_traits>
+
+#include "common/mixin/not_copyable.h"
+#include "runtime-core/runtime-core.h"
+#include "runtime-core/utils/kphp-assert-core.h"
+#include "runtime-light/utils/concepts.h"
+
+namespace tl {
+constexpr auto SMALL_STRING_SIZE_LEN = 1;
+constexpr auto MEDIUM_STRING_SIZE_LEN = 3;
+constexpr auto LARGE_STRING_SIZE_LEN = 7;
+
+constexpr uint64_t SMALL_STRING_MAX_LEN = 253;
+constexpr uint64_t MEDIUM_STRING_MAX_LEN = (static_cast<uint64_t>(1) << 24) - 1;
+[[maybe_unused]] constexpr uint64_t LARGE_STRING_MAX_LEN = (static_cast<uint64_t>(1) << 56) - 1;
+
+constexpr uint8_t LARGE_STRING_MAGIC = 0xff;
+constexpr uint8_t MEDIUM_STRING_MAGIC = 0xfe;
+
+class TLBuffer : private vk::not_copyable {
+  string_buffer m_buffer;
+  size_t m_pos{0};
+  size_t m_remaining{0};
+
+public:
+  TLBuffer() = default;
+
+  const char *data() const noexcept {
+    return m_buffer.buffer();
+  }
+
+  size_t size() const noexcept {
+    return static_cast<size_t>(m_buffer.size());
+  }
+
+  size_t pos() const noexcept {
+    return m_pos;
+  }
+
+  size_t remaining() const noexcept {
+    return m_remaining;
+  }
+
+  void clean() noexcept {
+    m_buffer.clean();
+    m_pos = 0;
+    m_remaining = 0;
+  }
+
+  void reset(size_t pos) noexcept {
+    php_assert(pos >= 0 && pos <= size());
+    m_pos = pos;
+    m_remaining = size() - m_pos;
+  }
+
+  void adjust(size_t len) noexcept {
+    php_assert(m_pos + len <= size());
+    m_pos += len;
+    m_remaining -= len;
+  }
+
+  void store_bytes(const char *src, size_t len) noexcept {
+    m_buffer.append(src, len);
+    m_remaining += len;
+  }
+
+  template<standard_layout T, standard_layout U>
+    requires std::convertible_to<U, T>
+  void store_trivial(const U &t) noexcept {
+    // Here we rely on that endianness of architecture is Little Endian
+    store_bytes(reinterpret_cast<const char *>(std::addressof(t)), sizeof(T));
+  }
+
+  void store_string(const char *str_buf, size_t str_len) noexcept;
```

Comment:
What do you think about changing the signature: replace these two parameters with a single `std::string_view`?